File: remedy/lib/pins.py
"""
MIDICaptain Remedy - Hardware Pin Definitions

Pin assignments for the Paint Audio MIDI Captain (RP2040-based)
Based on reverse engineering documentation.

Hardware:
- Microcontroller: Raspberry Pi Pico (RP2040)
- Display: ST7789 240x240 TFT LCD
- LEDs: 30 NeoPixels (3 per footswitch, 10 switches)
- Switches: 10 footswitches + 1 encoder button
- Encoder: Rotary with push button
- Expression: 2 analog inputs
- MIDI: UART TX/RX (5-pin DIN) + USB-MIDI
"""

import board

# ═══════════════════════════════════════════════════════════════════════════════
# FOOTSWITCHES
# ═══════════════════════════════════════════════════════════════════════════════
# Active LOW (pull-up, pressed = False)

# Numbered row (bottom)
SWITCH_1 = board.GP1      # Also used for boot mode detection
SWITCH_2 = board.GP25
SWITCH_3 = board.GP24
SWITCH_4 = board.GP23

# Lettered row (top)
SWITCH_A = board.GP9
SWITCH_B = board.GP10
SWITCH_C = board.GP11
SWITCH_D = board.GP18

# Navigation
SWITCH_UP = board.GP20
SWITCH_DOWN = board.GP19

# All footswitches in order (for iteration)
FOOTSWITCHES = [
    ('1', SWITCH_1),
    ('2', SWITCH_2),
    ('3', SWITCH_3),
    ('4', SWITCH_4),
    ('A', SWITCH_A),
    ('B', SWITCH_B),
    ('C', SWITCH_C),
    ('D', SWITCH_D),
    ('up', SWITCH_UP),
    ('down', SWITCH_DOWN),
]

# ═══════════════════════════════════════════════════════════════════════════════
# ROTARY ENCODER
# ═══════════════════════════════════════════════════════════════════════════════

ENCODER_A = board.GP2       # Phase A
ENCODER_B = board.GP3       # Phase B
ENCODER_SW = board.GP0      # Push button (active LOW)

# ═══════════════════════════════════════════════════════════════════════════════
# NEOPIXEL LEDs
# ═══════════════════════════════════════════════════════════════════════════════

NEOPIXEL_PIN = board.GP7
NEOPIXEL_COUNT = 30         # 3 LEDs per footswitch × 10 switches

# LED indices for each footswitch (3 LEDs per switch)
# NeoPixel chain order: 1,2,3,4 (top row) then up,A,B,C,D,down (bottom row)
LED_MAP = {
    '1': (0, 1, 2),
    '2': (3, 4, 5),
    '3': (6, 7, 8),
    '4': (9, 10, 11),
    'up': (12, 13, 14),
    'A': (15, 16, 17),
    'B': (18, 19, 20),
    'C': (21, 22, 23),
    'D': (24, 25, 26),
    'down': (27, 28, 29),
}

# ═══════════════════════════════════════════════════════════════════════════════
# DISPLAY (ST7789 240x240)
# ═══════════════════════════════════════════════════════════════════════════════

# SPI bus
SPI_CLK = board.GP14
SPI_MOSI = board.GP15
# SPI_MISO not used (display is write-only)

# Display control
TFT_CS = board.GP13         # Chip select
TFT_DC = board.GP12         # Data/Command
TFT_PWM = board.GP8         # Backlight PWM

# Display parameters
DISPLAY_WIDTH = 240
DISPLAY_HEIGHT = 240
DISPLAY_ROTATION = 180      # Mounted upside down
SPI_BAUDRATE = 24_000_000   # 24 MHz

# ═══════════════════════════════════════════════════════════════════════════════
# MIDI UART
# ═══════════════════════════════════════════════════════════════════════════════

MIDI_TX = board.GP16
MIDI_RX = board.GP17
MIDI_BAUDRATE = 31250       # Standard MIDI baud rate

# ═══════════════════════════════════════════════════════════════════════════════
# ANALOG INPUTS
# ═══════════════════════════════════════════════════════════════════════════════

EXPRESSION_1 = board.GP27   # A1 - Expression pedal 1
EXPRESSION_2 = board.GP28   # A2 - Expression pedal 2

# Battery voltage (GP29/A3) - not exposed on all CircuitPython builds
try:
    BATTERY_VOLTAGE = board.GP29
except AttributeError:
    BATTERY_VOLTAGE = None

# ADC reference voltage
ADC_REF_VOLTAGE = 3.3

# ═══════════════════════════════════════════════════════════════════════════════
# UNUSED/RESERVED GPIO
# ═══════════════════════════════════════════════════════════════════════════════
# These pins are available for future expansion:
# GP4, GP5, GP6, GP21, GP22, GP26

File: firmware/src/pins.rs
//! Pin map for the Paint Audio MIDI Captain.
//!
//! Mirrors `remedy/lib/pins.py` from the CircuitPython reference firmware.
//! When the OEM PCB silkscreens diverge from the reverse-engineered map,
//! update this file (and `HARDWARE.md`) — every other module routes
//! through here.
//!
//! Conventions:
//! - Footswitches: active LOW with internal pull-up. `Level::Low` means
//!   "pressed". The CP firmware drove these with `digitalio.Pull.UP` and
//!   read `False` on press; Embassy's `Input` with `Pull::Up` is identical.
//! - Display is mounted upside-down in the chassis. Initialize the
//!   `mipidsi` driver with `Orientation::Portrait(true).flip_horizontal()`
//!   or rotate by 180° when the display module lands.
//!
//! This module deliberately holds only constants and small type aliases —
//! no peripheral construction. Examples take `embassy_rp::init(...)` and
//! reach for `Peripherals::PIN_xx` directly, using these constants as
//! pin-number documentation. A later refactor may introduce
//! `assign-resources!`-style typed bundles per task.

#![allow(dead_code)]

// ── Footswitches ───────────────────────────────────────────────────────
// "Numbered" bottom row. SW1 doubles as the boot-mode detect pin in the
// CP firmware; with our own bootloader we may repurpose this — keep it as
// an input for now.
pub const SW_1_PIN:    u8 = 1;
pub const SW_2_PIN:    u8 = 25;
pub const SW_3_PIN:    u8 = 24;
pub const SW_4_PIN:    u8 = 23;

// "Lettered" top row.
pub const SW_A_PIN:    u8 = 9;
pub const SW_B_PIN:    u8 = 10;
pub const SW_C_PIN:    u8 = 11;
pub const SW_D_PIN:    u8 = 18;

// Navigation pair.
pub const SW_UP_PIN:   u8 = 20;
pub const SW_DOWN_PIN: u8 = 19;

/// Total physical footswitches (excluding the encoder push-button).
pub const FOOTSWITCH_COUNT: usize = 10;

// ── Rotary encoder ─────────────────────────────────────────────────────
pub const ENCODER_A_PIN:  u8 = 2;  // Quadrature phase A
pub const ENCODER_B_PIN:  u8 = 3;  // Quadrature phase B
pub const ENCODER_SW_PIN: u8 = 0;  // Push-button (active LOW)

// ── NeoPixels (WS2812B, single chain) ──────────────────────────────────
pub const NEOPIXEL_PIN:   u8    = 7;
pub const NEOPIXEL_COUNT: usize = 30; // 10 switches × 3 LEDs each

/// Switch label (parallel array index for [`LED_RANGES`] and the typed
/// pin map at the bottom of this module).
///
/// Order is the order the LEDs appear in the WS2812 daisy chain, NOT the
/// physical layout on the chassis. See `remedy/lib/pins.py::LED_MAP`.
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Switch {
    S1, S2, S3, S4,
    Up,
    A, B, C, D,
    Down,
}

impl Switch {
    pub const COUNT: usize = 10;

    /// Iterator-friendly array in chain order.
    pub const ALL: [Switch; Self::COUNT] = [
        Switch::S1, Switch::S2, Switch::S3, Switch::S4,
        Switch::Up,
        Switch::A,  Switch::B,  Switch::C,  Switch::D,
        Switch::Down,
    ];

    /// Index into the NeoPixel chain for the first LED of this switch.
    /// Each switch owns three contiguous LEDs.
    pub const fn led_start(self) -> usize {
        (self as usize) * 3
    }

    /// (start, count) range of NeoPixel indices for this switch.
    pub const fn led_range(self) -> (usize, usize) {
        (self.led_start(), 3)
    }
}

/// LED-index ranges parallel to [`Switch::ALL`], in chain order.
/// Equivalent to the `LED_MAP` dict in the CP reference.
pub const LED_RANGES: [(usize, usize); Switch::COUNT] = [
    (0, 3),   // S1
    (3, 3),   // S2
    (6, 3),   // S3
    (9, 3),   // S4
    (12, 3),  // Up
    (15, 3),  // A
    (18, 3),  // B
    (21, 3),  // C
    (24, 3),  // D
    (27, 3),  // Down
];

// ── ST7789 240×240 TFT ────────────────────────────────────────────────
// On RP2040, GP14/GP15 are SPI1 SCK/TX. CS/DC are plain GPIO outputs.
// Backlight is PWM-driven (CP used pwmio at default freq); plain GPIO
// works as well if you don't need dimming.
pub const SPI_CLK_PIN:    u8 = 14;  // SPI1 SCK
pub const SPI_MOSI_PIN:   u8 = 15;  // SPI1 TX  (display is write-only)
pub const TFT_CS_PIN:     u8 = 13;
pub const TFT_DC_PIN:     u8 = 12;
pub const TFT_BACKLIGHT_PIN: u8 = 8; // PWM-capable

pub const DISPLAY_WIDTH:    u16  = 240;
pub const DISPLAY_HEIGHT:   u16  = 240;
// Panel is physically chassis-inverted, but mipidsi 0.10's rotation/offset
// convention differs from CircuitPython's: verified on hardware that
// Rotation::Deg0 + display_offset(0,0) reads upright and centred. (CP's
// adafruit_st7789 uses rotation=180 + rowstart=80 for the same result — do
// not copy those numbers into mipidsi.) See display.rs module header.
pub const DISPLAY_ROTATION_DEG: u16 = 0;
pub const DISPLAY_SPI_BAUD: u32  = 24_000_000;

// ── DIN MIDI (5-pin) over UART0 ────────────────────────────────────────
// GP16 = UART0 TX, GP17 = UART0 RX (RP2040 alt-function 2). Standard
// MIDI baud is exactly 31250.
pub const MIDI_TX_PIN: u8  = 16;
pub const MIDI_RX_PIN: u8  = 17;
pub const MIDI_BAUD:   u32 = 31_250;

// ── Expression pedals (TRS, tip = wiper) ──────────────────────────────
// ADC1 (GP27) and ADC2 (GP28). Calibration min/max live in flash; the CP
// firmware put them in NVM because its filesystem was read-only.
pub const EXPRESSION_1_PIN: u8 = 27;
pub const EXPRESSION_2_PIN: u8 = 28;

// Optional battery-voltage divider on ADC3 (GP29). Present on some
// boards, absent on others — code that uses this must tolerate it being
// unimplemented.
pub const BATTERY_VOLTAGE_PIN: u8 = 29;

/// ADC reference (RP2040 internal Vref ≈ 3.3 V).
pub const ADC_REF_VOLTAGE: f32 = 3.3;

// ── Reserved / unused GPIOs ───────────────────────────────────────────
// GP4, GP5, GP6, GP21, GP22, GP26. Available for future expansion.

// ── USB device identity ───────────────────────────────────────────────
// The Raspberry Pi vendor ID. KEEP THIS — scripts/bootsel_hammer.py
// matches devices by VID, and any change here breaks the recovery flow.
pub const USB_VID: u16 = 0x2E8A;
/// Development-time PID. Pick a value outside the Pico SDK's well-known
/// range to avoid confusing host udev rules / drivers. Update when this
/// firmware ships under a stable identity.
pub const USB_PID: u16 = 0x102D;
pub const USB_MANUFACTURER: &str = "Paint Audio";
pub const USB_PRODUCT:      &str = "MIDICaptain Remedy (Rust)";
